Determine P-site offsets for :term:`ribosome profiling` data
============================================================

In this tutorial we determine the :term:`P-site offset` to use with a
:term:`ribosome profiling` dataset. We will estimate a separate
offset for each read length, as footprints of different lengths
might require different offsets.

We can then use these offsets as a :term:`mapping rule` to assign
P-sites to :term:`read alignments` for future positional analysis of
:term:`ribosome profiling` data.

Here, we presume that ribosome-protected fragments were produced using
an unbiased nuclease (e.g. RNase I, NOT microccal nuclease). that
trims the :term:`ribosome-protected footprints <footprint>` to the
edges of the ribosome.

We will use the :doc:`/test_dataset` of yeast chromosome I.


Strategy
--------
Following :cite:`Ingolia2009` and others, an effective way to map P-sites
is to use peaks of ribosome density that occur over start codons.

Because the ribosomal P-site is somewhere internal to the
:term:`ribosome protected footprint` generated by ribosome profiling, when
we map :term:`read alignments` of :term:`footprints <footprint>` to their
5' ends, we would expect to see a peak of ribosome density *upstream* of
the start codon. The distance between the peak and the start codon
corresponds to the offset.

This yields the following strategy:

 #. Separate :term:`footprints <footprint>` into classes based upon their lengths

 #. For each length:

      #. Perform a :term:`metagene analysis` around the start codon,
         in which the :term:`footprints <footprint>` are mapped to their 5' ends.

      #. Measure the distance between the highest peak 5' of the start codon
         and the start codon. Assuming this peak is the initiation peak, this
         distance is the offset to use for reads of this length.

 #. Manually inspect offsets to make sure they seem reasonable

 #. Check results by perform a :term:`metagene analysis` around the start codon, 
    this time using the :term:`P-site offsets <P-site offset>` determined above.


Determining :term:`P-site offsets <P-site offset>` using the |psite| script
---------------------------------------------------------------------------
The strategy above is implemented by |psite|, which can be
executed from the terminal.

Because |psite| internally performs :term:`metagene analysis`, we need
to use a file produced by the |metagene| script. The command call to 
|metagene| is included below, and explained in detal in :doc:`/examples/metagene`.
From the terminal:

 .. code-block:: shell

    # generate metagene `roi` file. We only need to do this once,
    # or when we switch to a new genome annotation.
    #
    # this will make 'chrI_rois.txt' and 'chrI_rois.bed'. They have the same
    # information, but the BED file can be loaded into a genome browser
    $ metagene generate chrI --landmark cds_start --annotation_files sgd_plus_utrs_chrI.gtf

    # run the psite script
    # We ignore reads shorter than 25 nucleotides or longer than 35-
    # there should be few of these, and it saves psite from doing 
    # unnecessary analyses
    $ psite chrI_rois.txt SRR1562907 --min_length 25 --max_length 35 --require_upstream --count_files SRR1562907_chrI.bam

The script will make many files, two of which are of interest:

  #. A two-column text file (``SRR1562907_p_offsets.txt``), in which the first column is a read length and the
     second, the corresponding P-site offset from the 5' end of the read.
     Comments and metadata appear up top. In the example above, the file looks
     like::

        ## date = '2015-07-16 13:51:29.984305'
        ## args = {  
        ##          'count_files'      : ['SRR1562907_chrI.bam'],
        ##          'countfile_format' : 'BAM',
        ##          'default'          : 13,
        ##          'mapping'          : 'fiveprime',
        ##          'max_length'       : 35,
        ##          'min_counts'       : 10,
        ##          'min_length'       : 25,
        ##          'nibble'           : 0,
        ##          'norm_region'      : (70, 100),
        ##          'offset'           : 0,
        ##          'outbase'          : 'SRR1562907',
        ##          'require_upstream' : True,
        ##          'roi_file'         : 'chrI_rois.txt'
        ##        }
        #length	p_offset
        25	9
        26	12
        27	11
        28	12
        29	12
        30	13
        31	13
        32	13
        33	13
        34	13
        35	13
        default	13

  #. An SVG graphic (``SRR1562907_p_offsets.svg``), showing the metagene
     profiles for each read length:

        [TODO: include updated graphic]


Manually determining :term:`P-site offsets <P-site offset>`
-----------------------------------------------------------
Those interested in manually performing P-site offset analysis should refer to 
the source code, first for the |metagene| ``count`` subprogram, and then for |psite|.


Using the P-site offset in analyses
-----------------------------------

The goal of this analysis is to determine a :term:`P-site offset` for use
as a :term:`mapping rule` in subsequent analysis of a dataset.


In command-line scripts
.......................

Command-line scripts in :mod:`yeti <yeti.bin>` use a common interface for
read :term:`mapping rules <mapping rule>`. To use the offsets generated by |psite|, use
the ``--fiveprime_variable`` mapping rule, and pass the text file made
by |psite| to the ``--offset`` parameter. For example, from the terminal:

 .. code-block :: shell

    $ some_script --offset SRR1562907_p_offsets.txt --fiveprime_variable ... [other arguments]


In interactive sessions
.......................

In interactive sessions, we first need to load the offset file::

    >>> offset_dict = {}

    >>> with open("SRR1562907_p_offsets.txt") as fin:
    >>>     for line in fin: 
    >>>         if not line.startswith("#"): # ignore comments & metadata
    >>>             length, offset = line.strip("\n").split("\t")
    >>>             offset_dict[length] = int(offset)


And then pass it to the appropriate mapping rule. For alignments in `BAM`_
format, use |BAMGenomeArray|::

    >>> import pysam
    >>> from yeti.genomics.genome_array import BAMGenomeArray, VariableFivePrimeMapFactory
    
    >>> alignments = BAMGenomeArray([pysam.Samfile("SRR1562907.bam","rb")])
    >>> alignments.set_mapping(VariableFivePrimeMapFactory(offset_dict))


.. TODO : create bowtie file?

For alignments in `bowtie`_-format use |GenomeArray|::

    >>> from yeti.genomics.genome_array import GenomeArray, variable_five_prime_map

    >>> alignments = GenomeArray()
    >>> alignments.add_from_bowtie("some_file.bowtie",variable_five_prime_map ,offset=offset_dict)


Pitfalls
--------

This P-site mapping strategy requires pronounced initiation peaks in
:term:`ribosome profiling` data. If these are absent -- which can
happen under conditions of initiation shutdown (if the sample is under
stress before lysis) -- an alternative option is to use a stop codon
peak (if present) for mapping.

The simplest way to do this is to use the |metagene| script on reads
of separate lengths, again using fiveprime end mapping (``--fiveprime``
command-line argument passed to |metagene|), and manually inspecting
the output. For each read length, assign the offset to be the distance
between the stop codon and the peak (which should be tall, and followed
by a precipitous drop in ribosome density) immediately upstream of
the stop codon.


Is it necessary to do this separately for every dataset?
--------------------------------------------------------
Many experimentalists find that their technique is sufficiently consistent
not to need to re-estimate P-site offsets for every dataset. Others are
content to use offsets published in literature by other groups. Others
more conservatively perform this analysis for every dataset. We strongly
suggest performing this analysis at the very least:

  - when changing nuclease, buffer, or cloning conditions

  - when changing culture conditions (e.g. profiling under starvation,
    heat shock, viral infection, et c)


See also
--------

  - |psite| script

  - |metagene| script

  - :doc:`/examples/metagene`
